### PR TITLE
add tip spot generators

### PR DESCRIPTION
**Pull request by rickwierenga**
in a new `resources.functional` module, similar to what PyTorch has

```python
import pylabrobot.resources.functional as F
tip_racks = [tr0, tr1, ...]
gen = F.linear_tip_spot_generator(tip_racks, cache_file_path="my_cache.json")

...

num_tips_needed = 3
tip_spots = [await gen.__anext__() for _ in range(num_tips_needed)]
await lh.pick_up_tips(tip_spots)
````

### CODE REVIEW

**Comment by rickwierenga** [comment]
todo: also accept `List[TipSpot]`

File: pylabrobot/resources/functional.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+async def linear_tip_spot_generator(tip_racks: List[TipRack],
+                                    cache_file_path: Optional[str]=None):
+  """ Tip spot generator with disk caching. Linearly iterate through all tip spots and
+  raise StopIteration when all spots have been used. """
+  tip_rack_idx, tip_spot_idx = 0, 0
+  if cache_file_path is not None and os.path.exists(cache_file_path):
+    with open(cache_file_path, "r", encoding="utf-8") as f:
+      data = json.load(f)
+      tip_rack_idx, tip_spot_idx = data["tri"], data["tsi"]
+      print("loaded tip idx from disk:", data)
```
Comment (jkhales) [inline pylabrobot/resources/functional.py:18]: switch to logger?

File: pylabrobot/resources/functional.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  while tip_rack_idx < len(tip_racks):
+    while tip_spot_idx < 96:
+      tip_spot_idx += 1
+      if cache_file_path is not None:
+        with open(cache_file_path, "w", encoding="utf-8") as f:
+          json.dump({"tri": tip_rack_idx, "tsi": tip_spot_idx}, f)
```
Comment (jkhales) [inline pylabrobot/resources/functional.py:25]: why not just spell out `tip_rack_idx` and `tip_spot_idx`? abbreviations are the enemy of readability

File: pylabrobot/resources/functional.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                                        cache_file_path: Optional[str]=None):
+  """ Randomized tip spot generator with disk caching. Don't return tip spots that have been
+  sampled in the last K samples. """
+  
+  all_tip_spots = [rack.get_item(spot_idx) for rack in tip_racks for spot_idx in range(rack.num_items)]
+  recently_sampled = deque(maxlen=K)
+
+  if cache_file_path is not None and os.path.exists(cache_file_path):
+    with open(cache_file_path, "r", encoding="utf-8") as f:
+      data = json.load(f)
```
Comment (jkhales) [inline pylabrobot/resources/functional.py:42]: have you considered converting this into a dataclass? this allows readers of the code to see what was intended and validates input

File: pylabrobot/resources/functional.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        with open(cache_file_path, "w", encoding="utf-8") as f:
+          json.dump({"tri": tip_rack_idx, "tsi": tip_spot_idx}, f)
+      yield tip_racks[tip_rack_idx][tip_spot_idx]
+
+    tip_rack_idx += 1
+    tip_spot_idx = 0
+
+
+async def randomized_tip_spot_generator(tip_racks: List[TipRack], K: int,
+                                        cache_file_path: Optional[str]=None):
```
Comment (jkhales) [inline pylabrobot/resources/functional.py:33]: return type?

File: pylabrobot/resources/functional.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import os
+import json
+import random
+from collections import deque
+from typing import List, Optional
+from pylabrobot.resources.tip_rack import TipRack
+
+
+async def linear_tip_spot_generator(tip_racks: List[TipRack],
+                                    cache_file_path: Optional[str]=None):
```
Comment (jkhales) [inline pylabrobot/resources/functional.py:10]: return type?

File: pylabrobot/resources/functional.py (hunk 2 above)
Comment (rickwierenga) [reply pylabrobot/resources/functional.py:25]: resource style code artifacts 😅

File: pylabrobot/resources/functional.py (hunk 3 above)
Comment (rickwierenga) [reply pylabrobot/resources/functional.py:42]: what would the dataclass be in this case? it's a single parameter

Comment (rickwierenga) [reply pylabrobot/resources/functional.py:42]: for storing the data on disk?

Comment (jkhales) [reply pylabrobot/resources/functional.py:49]: for bringing it back from disk.

this ensure the json wasn't corrupted and allows people to see what it looks like. 

It's a nice to have, not a must have.

Comment (rickwierenga) [reply pylabrobot/resources/functional.py:49]: probably in a future version

**Comment by rickwierenga** [comment]
final syntax now that it takes tip spots:

```python
import pylabrobot.resources.functional as F
tip_spots = F.get_all_tip_spots([tiprack0, tiprack1, ...])
gen = F.linear_tip_spot_generator(tip_racks, cache_file_path="my_cache.json")

...

num_tips_needed = 3
tip_spots = [await gen.__anext__() for _ in range(num_tips_needed)]
await lh.pick_up_tips(tip_spots)
```

**Pull request merged**